cuequivariance_torch/cuequivariance_torch/primitives/segmented_polynomial_fused_tp.py

```
@@ -0,0 +1,233 @@
+
+            if d.num_operands not in (3, 4):
+                raise NotImplementedError(
+                    f"Fused TP only supports 3 or 4 operands. Got {d.subscripts}."
+                )
+
+            try:
+                d, perm = next(
+                    cue.segmented_polynomials.dispatch(
+                        d, supported_targets, "permute_all_but_last"
+                    )
+                )
+            except StopIteration:
+                raise NotImplementedError(
+                    f"Fused TP does not support {d}."
+                    " Supported targets are: "
+                    + ", ".join(str(t) for t in supported_targets)
+                )
+
+            if d.num_operands == 3:
+                self.tps.append(FusedTP3(d, self.math_dtype))
+            elif d.num_operands == 4:
+                self.tps.append(FusedTP4(d, self.math_dtype))
+
+            self.input_inds.append(
+                [perm[i] for i in operation.input_buffers(self.num_inputs)]
+            )
+            self.b_outs.append(b_out - self.num_inputs)
+
+    def forward(
+        self,
+        inputs: List[torch.Tensor],
+        input_indices: Dict[int, torch.Tensor],
+        output_shapes: Dict[int, torch.Tensor],
+        output_indices: Dict[int, torch.Tensor],
+    ):
+        for i, x, size in zip(range(self.num_inputs), inputs, self.input_sizes):
+            if x.ndim == 0:
+                raise ValueError(f"Input {i} has no dimensions")
+            if x.shape[-1] != size:
+                raise ValueError(
+                    f"Input {i} has shape {x.shape} but expected shape {size}."
+                )
+
+        # This is not supported:
+        if self.math_dtype == torch.float32 and any(
+            t.dtype == torch.float64 for t in inputs
+        ):
+            raise ValueError(
+                "Fused TP does not support float32 math_dtype with float64 inputs."
+            )
```

can we add here the "dynamic" math_dtype for when it's set to None? Otherwise if we cant have a dynamic math_dtype let's raise an error and ask the user to specify a math_dtype

---

Yeah, the issue here is that with FusedTP I need to define the dtype at initialization. Maybe that can be worked around but I didn't see a quick way. But you're right: I'll raise an error here.

---

Btw, I'm trying to capture this in the docstring of the base class. I'll make sure everything is aligned.